<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Julien Racle and David Gfeller" />

<meta name="date" content="2023-07-12" />

<title>EPIC package</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">EPIC package</h1>
<h4 class="author">Julien Racle and David Gfeller</h4>
<h4 class="date">2023-07-12</h4>



<div id="description" class="section level2">
<h2>Description</h2>
<p>Package implementing EPIC method to estimate the proportion of
immune, stromal, endothelial and cancer or other cells from bulk gene
expression data. It is based on reference gene expression profiles for
the main non-malignant cell types and it predicts the proportion of
these cells and of the remaining “other cells” (that are mostly cancer
cells) for which no reference profile is given.</p>
<p>This method is described in the publication from <em>Racle et al.,
2017</em> available at <a href="https://elifesciences.org/articles/26476" class="uri">https://elifesciences.org/articles/26476</a>.</p>
<p>EPIC is also available as a web application: <a href="http://epic.gfellerlab.org" class="uri">http://epic.gfellerlab.org</a>.</p>
</div>
<div id="usage" class="section level2">
<h2>Usage</h2>
<p>The main function in this package is <code>EPIC</code>. It needs as
input a matrix of the TPM (or RPKM) gene expression from the samples for
which to estimate cell proportions. One can also define the reference
cells to use</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># library(EPIC) ## If the package isn&#39;t loaded (or use EPIC::EPIC and so on).</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>out <span class="ot">&lt;-</span> <span class="fu">EPIC</span>(<span class="at">bulk =</span> bulkSamplesMatrix)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>out <span class="ot">&lt;-</span> <span class="fu">EPIC</span>(<span class="at">bulk =</span> bulkSamplesMatrix, <span class="at">reference =</span> referenceCellsList)</span></code></pre></div>
<p><code>out</code> is a list containing the various mRNA and cell
fractions in each samples as well as some <em>data.frame</em> of the
goodness of fit.</p>
<p>Values of mRNA per cell and signature genes to use can also be
changed:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>out <span class="ot">&lt;-</span> <span class="fu">EPIC</span>(<span class="at">bulk =</span> bulkSamplesMatrix, <span class="at">reference =</span> referenceCellsList, <span class="at">mRNA_cell =</span> mRNA_cell_vector, <span class="at">sigGenes =</span> sigGenes_vector)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>out <span class="ot">&lt;-</span> <span class="fu">EPIC</span>(<span class="at">bulk =</span> bulkSamplesMatrix, <span class="at">reference =</span> referenceCellsList, <span class="at">mRNA_cell_sub =</span> mRNA_cell_sub_vector)</span></code></pre></div>
<p>Various other options are available and are well documented in the
help pages from EPIC:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>?EPIC<span class="sc">::</span>EPIC</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>?EPIC<span class="sc">::</span>EPIC.package</span></code></pre></div>
</div>
<div id="installation" class="section level2">
<h2>Installation</h2>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;devtools&quot;</span>)</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">&quot;GfellerLab/EPIC&quot;</span>, <span class="at">build_vignettes=</span><span class="cn">TRUE</span>)</span></code></pre></div>
</div>
<div id="web-application" class="section level2">
<h2>Web application</h2>
<p>EPIC is also available as a web application: <a href="http://epic.gfellerlab.org" class="uri">http://epic.gfellerlab.org</a>.</p>
</div>
<div id="python-wrapper" class="section level2">
<h2>Python wrapper</h2>
<p>A pyhton wrapper has been written by Stephen C. Van Nostrand from MIT
and is available at <a href="https://github.com/scvannost/epicpy" class="uri">https://github.com/scvannost/epicpy</a>.</p>
</div>
<div id="license" class="section level2">
<h2>License</h2>
<p>EPIC can be used freely by academic groups for non-commercial
purposes. The product is provided free of charge, and, therefore, on an
“<em>as is</em>” basis, without warranty of any kind. Please read the
file “<em>LICENSE</em>” for details.</p>
<p>If you plan to use EPIC (version 1.1) in any for-profit application,
you are required to obtain a separate license. To do so, please contact
Nadette Bulgin (<a href="mailto:nbulgin@lcr.org">nbulgin@lcr.org</a>) at
the Ludwig Institute for Cancer Research Ltd.</p>
</div>
<div id="contact-information" class="section level2">
<h2>Contact information</h2>
<p>Julien Racle (<a href="mailto:julien.racle@unil.ch">julien.racle@unil.ch</a>), and David
Gfeller (<a href="mailto:david.gfeller@unil.ch">david.gfeller@unil.ch</a>).</p>
</div>
<div id="faq" class="section level2">
<h2>FAQ</h2>
<div id="which-proportions-returned-by-epic-should-i-use" class="section level5">
<h5>Which proportions returned by EPIC should I use?</h5>
<ul>
<li><p>EPIC is returning two proportion values: <em>mRNAProportions</em>
and <em>cellFractions</em>, where the 2nd represents the true proportion
of cells coming from the different cell types when considering
differences in mRNA expression between cell types. So in principle, it
is best to consider these <em>cellFractions</em>.</p>
<p>However, please note, that when the goal is to benchmark EPIC
predictions, if the ‘bulk samples’ correspond in fact to in silico
samples reconstructed for example from single-cell RNA-seq data, then it
is usually better to compare the ‘true’ proportions against the
<em>mRNAProportions</em> from EPIC. Indeed, when building such in silico
samples, the fact that different cell types express different amount of
mRNA is usually not taken into account. On the other side, if working
with true bulk samples, then you should compare the true cell
proportions (measured e.g., by FACS) against the
<em>cellFractions</em>.</p></li>
</ul>
</div>
<div id="what-do-the-other-cells-represent" class="section level5">
<h5>What do the “<em>other cells</em>” represent?</h5>
<ul>
<li>EPIC predicts the proportions of the various cell types for which we
have gene expression reference profiles (and corresponding gene
signatures). But, depending on the bulk sample, it is possible that some
other cell types are present for which we don’t have any reference
profile. EPIC returns the proportion of these remaining cells under the
name “<em>other cells</em>”. In the case of tumor samples, most of these
other cells would certainly correspond to the cancer cells, but it could
be that there are also some stromal cells or epithelial cells for
example.</li>
</ul>
</div>
<div id="i-receive-an-error-message-attempt-to-set-colnames-on-an-object-with-less-than-two-dimensions.-what-can-i-do" class="section level5">
<h5>I receive an error message “<em>attempt to set ‘colnames’ on an
object with less than two dimensions</em>”. What can I do?</h5>
<ul>
<li>This is certainly that some of your data is a vector instead of a
matrix. Please make sure that your bulk data is in the form of a matrix
(and also your reference gene expression profiles if using custom
ones).</li>
</ul>
</div>
<div id="is-there-some-caution-to-consider-about-the-cellfractions-and-mrna_cell-values" class="section level5">
<h5>Is there some caution to consider about the <em>cellFractions</em>
and <em>mRNA_cell</em> values?</h5>
<ul>
<li><p>As described in our manuscript, EPIC first estimates the
proportion of mRNA per cell type in the bulk and then it uses the fact
that some cell types have more mRNA copies per cell than other to
normalize this and obtain an estimate of the proportion of cells instead
of mRNA (EPIC function returns both information if you need the one or
the other). For this normalization we had either measured the amount of
mRNA per cell or found it in the literature (fig. 1 – fig. supplement 2
of our paper). However we don’t currently have such values for the
endothelial cells and CAFs. Therefore for these two cell types, we use
an average value, which might not reflect their true value and this
could bias a bit the predictions, especially for these cell types. If
you have some values for these mRNA/cell abundances, you can also add
them into EPIC, with help of the parameter “<em>mRNA_cell</em>” or
“<em>mRNA_cell_sub</em>” (and that would be great to share these
values).</p>
<p>If the mRNA proportions of these cell types are low, then even if you
don’t correct the results with their true mRNA/cell abundances, it would
not really have a big impact on the results. On the other side, if there
are many of these cells in your bulk sample, the results might be a
little bit biased, but the effect should be similar for all samples and
thus not have a too big importance (maybe you wouldn’t be fully able to
tell if there are more CAFs than Tcells for example, but you should
still have a good estimate of which sample has more CAFs (or Tcells)
than which other sample for example).</p></li>
</ul>
</div>
<div id="i-receive-a-warning-message-that-the-optimization-didnt-fully-converge-for-some-samples.-what-does-it-mean" class="section level5">
<h5>I receive a warning message that “<em>the optimization didn’t fully
converge for some samples</em>”. What does it mean?</h5>
<ul>
<li><p>When estimating the cell proportions EPIC performs a least square
regression between the observed expression of the signature genes and
the expression of these genes predicted based on the estimated
proportions and gene expression reference profiles of the various cell
types.</p>
<p>When such a warning message appears, it means that the optimization
didn’t manage to fully converge for this regression, for some of the
samples. You can then check the “<em>fit.gof$convergeCode</em>” (and
possibly also “<em>fit.gof$convergeMessage</em>”) that is outputted by
EPIC alongside the cell proportions. This will tell you which samples
had issue with the convergence (a value of 0 means it converged ok,
while other values are errors/warnings, their meaning can be found in
the help of “<em>optim</em>” (or “<em>constrOptim</em>”) function from R
(from “<em>stats</em>” package) which is used during the optimization
and we simply forward the message it returns).</p>
<p>The error code that usually comes is a “1” which means that the
maximum number of iterations has been reached in the optimization. This
could mean there is an issue with the bulk gene expression data that
maybe don’t completely follow the assumption of equation (1) from our
manuscript. From our experience, it seems in practice that even when
there was such a warning message the proportions were predicted well, it
is maybe that the optimization just wants to be <em>too precise</em>, or
maybe few of the signature genes didn’t match well but the rest of
signature genes could be used to have a good estimate of the
proportions.</p>
<p>If you have some samples that seem to have strange results, it could
however be useful to check that the issue is not that these samples
didn’t converge well. To be more conservative you could also remove all
the samples that didn’t converge well as these are maybe outliers, if it
is only a small fraction from your original samples. Another possibility
would be to change the parameters of the optim/constrOptim function to
allow for more iterations or maybe a weaker tolerance for the
convergence, but for this you would need to tweak it directly in the
code of EPIC, I didn’t implement such option for EPIC.</p></li>
</ul>
</div>
<div id="who-should-i-contact-in-case-of-a-technical-or-other-issue" class="section level5">
<h5>Who should I contact in case of a technical or other issue?</h5>
<ul>
<li>Julien Racle (<a href="mailto:julien.racle@unil.ch">julien.racle@unil.ch</a>). Please
provide as much details as possible and ideally send also an example
input file (and/or reference profiles) that is causing the issue.</li>
</ul>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
